Oracle.md

```diff
@@ -0,0 +1,160 @@
+# Oracle specifications
+
+## Introduction
+
+For the purpose of these specifications, an event is a digital representation of a real world fact.
+An [oracle](./Introduction#Oracle) is an entity that commits to publishing a signature over a (number of) event outcome(s) ahead of time by releasing an [R-value](./Introduction#R-value) as well as necessary information for two parties to build a set of [CETs](./Introduction#Contract-Execution-Transaction-(CET)).
+This necessary information is committed in a so-called _event descriptor_ that will be further detailed in this document.
+Descriptors are classified and discovered through a namespace presently represented as Unique Resource Identifier (URI) paths.
+
+## Event descriptor
+
+An event descriptor provides information to clients about an event for which an oracle plans on releasing a signature over its outcome.
+The provided information should be sufficient for a client to create a set of CETs that will cover all possible outcomes of the event.
+
+Here we assume that an event outcome can be represented either as a set of strings or numbers.
+Four kinds of event outcomes are defined (note that a single even can be composed of several types).
+
+### External event descriptors
+
+External event descriptor enables representation of oracles in a format agreed upon between an oracle and its clients, for the cases where the following descriptor types are inadequate.
+The structure and semantic of the information contained in the descriptor is thus open.
+
+### Simple enumeration
+
+For events that have a narrow range of possible outcomes, the outcomes can simply be enumerated.
+For the sake of simplicity, outcomes should be represented as strings.
+
+#### Example: Weather tomorrow
+
+Tomorrow's weather can be represented as the set of strings `[sunny, cloudy, rainy]`.
+
+### Range
+
+When an event has numerical outcomes that cannot be easily enumerated, they can be represented as a series with:
+
+- start: the first possible outcome number
+- count: the number of possible outcomes
+- step: the increment
+
+#### Example: tomorrow's temperature
+
+```
+start: -100
+count: 201
+step: 1
+```
+
+### Large range
+
+When a range of a numerical outcomes is large or unbounded, the oracle can represent the outcome using numerical decomposition.
+
+The event descriptor should include:
+
+- base: the base in which the outcome value is decomposed and which will indicate the possible range of each digit that will be signed.
+- r-values: an array of R values, one for each of the digit that will be signed by the oracle.
+
+In the case where the number of digits that make up the outcome value exceeds the number of r-values that the oracle committed to, the oracle should sign the maximum possible value that he can attest to. 
```

Comment:
There should be an extra R value for signing if the value is negative or positive

Reply:
Also do we want to add handling for overflow?

Reply:
> There should be an extra R value for signing if the value is negative or positive

Good point. I ended up adding a `is-signed` field because I feel it's a shame to always require an extra R-values but let me know what you think.

 > Also do we want to add handling for overflow?

There was quite some discussions around this (e.g. https://github.com/discreetlogcontracts/dlcspecs/pull/55#discussion_r487775283 or https://github.com/discreetlogcontracts/dlcspecs/issues/65) but the conclusion was that it's simplest for the oracle to just sign the highest possible value that it can given the number of R-values that it committed to. I think we should have some sort of event versioning so that an oracle can increase the number of R-values for a given event for new contracts while still be able to cater for old ones, but that's probably for the next iteration I would say.